README.rst
******************
eagerx_ode package
******************

.. image:: https://img.shields.io/badge/License-Apache_2.0-blue.svg
   :target: https://opensource.org/licenses/Apache-2.0
   :alt: license

.. image:: https://img.shields.io/badge/code%20style-black-000000.svg
   :target: https://github.com/psf/black
   :alt: codestyle

.. image:: https://github.com/eager-dev/eagerx_ode/actions/workflows/ci.yml/badge.svg?branch=master
   :target: https://github.com/eager-dev/eagerx_ode/actions/workflows/ci.yml
   :alt: Continuous Integration

.. contents:: Table of Contents
    :depth: 2

What is the *eagerx_ode* package?
=================================
This repository/package contains the *OdeEngine* for simulating robots based on ordinary differential equations (ODEs) in EAGERx.
EAGERx (Engine Agnostic Graph Environments for Robotics) enables users to easily define new tasks, switch from one sensor to another, and switch from simulation to reality with a single line of code by being invariant to the physics engine.

`The core repository is available here <https://github.com/eager-dev/eagerx>`_.

`Full documentation and tutorials (including package creation and contributing) are available here <https://eagerx.readthedocs.io>`_.

Installation
============

You can install the package using pip:

.. code:: shell

    pip3 install eagerx-ode

Cite EAGERx
===========
If you are using EAGERx for your scientific publications, please cite:

.. code:: bibtex

    @article{eagerx,
        author  = {van der Heijden, Bas and Luijkx, Jelle, and Ferranti, Laura and Kober, Jens and Babuska, Robert},
        title = {EAGERx: Engine Agnostic Graph Environments for Robotics},
        year = {2022},
        publisher = {GitHub},
        journal = {GitHub repository},
        howpublished = {\url{https://github.com/eager-dev/eagerx}}
    }

Acknowledgements
=================
EAGERx is funded by the `OpenDR <https://opendr.eu/>`_ Horizon 2020 project.


CHANGELOG.md
# Changelog

<!--next-version-placeholder-->

## v0.1.26 (2022-09-13)
### Fix
* Graph refactor ([#41](https://github.com/eager-dev/eagerx_ode/issues/41)) ([`e96ac66`](https://github.com/eager-dev/eagerx_ode/commit/e96ac668d9c84b083784fda9bb148910e450e20c))

## v0.1.25 (2022-07-20)


## v0.1.12 (2022-05-16)
### Fix
* Update dependencies ([`aebceaf`](https://github.com/eager-dev/eagerx_ode/commit/aebceaf3734a09aba9071f2eebaddf03b9c9eb4d))
* Refactor bridge to engine ([`5250d06`](https://github.com/eager-dev/eagerx_ode/commit/5250d0693801290d11a8a66ac944a01bcae92a0e))
* Refactor initialize_spec ([`e44c466`](https://github.com/eager-dev/eagerx_ode/commit/e44c466bace045945c9ab8ad5b400989f1305c06))

### Documentation
* Update  bibtex typo ([`fd5c2be`](https://github.com/eager-dev/eagerx_ode/commit/fd5c2bee05c4551df615db86db3435d1b3b3b076))

## v0.1.11 (2022-04-26)
### Fix
* Bump flake8, eagerx, pytest ([`1ffd80f`](https://github.com/eager-dev/eagerx_ode/commit/1ffd80f9e98028d68efb1b8f194c82a6f9acd5d5))
* Refactor `is_reactive` to `sync` ([`095741e`](https://github.com/eager-dev/eagerx_ode/commit/095741e7b6fb004a5f044cc1353114b031909716))

## v0.1.10 (2022-04-22)
### Fix
* Add engine node for float32 observations ([`e9127cc`](https://github.com/eager-dev/eagerx_ode/commit/e9127cce4657a3c55b0712e584ddd471e902fac9))

## v0.1.9 (2022-04-21)
### Fix
* Dfun odeint ([`ec3edf9`](https://github.com/eager-dev/eagerx_ode/commit/ec3edf9d59943e4b8badf3a9918d2e010f29bb3a))

## v0.1.8 (2022-04-20)
### Fix
* Use environment process per default ([`3283464`](https://github.com/eager-dev/eagerx_ode/commit/3283464756b315de7ef3875050b5a3e56630183c))

## v0.1.7 (2022-04-14)
### Fix
* All messages of observation and action_applied now inputs to render_fn ([`c279562`](https://github.com/eager-dev/eagerx_ode/commit/c279562d9bb75def98ef4709dd76a91a59e8fbaf))

## v0.1.6 (2022-04-14)
### Fix
* Add OdeRender EngineNode ([`f383d37`](https://github.com/eager-dev/eagerx_ode/commit/f383d372f45103f523c8d721a87ced8dd11594bf))

## v0.1.5 (2022-04-08)
### Fix
* Bump eagerx version to 0.1.14 ([`304a662`](https://github.com/eager-dev/eagerx_ode/commit/304a6621cdc89295d3e9633e050094b73ac9ffd9))
* Flake8 constraint ([`4fc27ca`](https://github.com/eager-dev/eagerx_ode/commit/4fc27cae88e2e50f0185d0d1aeba36dab94475d1))

## v0.1.4 (2022-03-23)
### Fix
* Api changes eagerx ([`1e64cb6`](https://github.com/eager-dev/eagerx_ode/commit/1e64cb64b75d1ce8f337145e7e6bdfca2a9476ce))

### Documentation
* **readme:** Add installation instructions ([`a9848ad`](https://github.com/eager-dev/eagerx_ode/commit/a9848adf5285875df27aa87b160f9ebec9211fc2))
* **README:** Fix typo ([`4bf5571`](https://github.com/eager-dev/eagerx_ode/commit/4bf557164b941c3b99a63cbbd41990672abca37f))
* Fix indentation README.rst ([`7df4ad6`](https://github.com/eager-dev/eagerx_ode/commit/7df4ad6b65b9eecb0db458d1f075bc428bfe046b))
* Add buttons for code coverage and ci ([`475a99f`](https://github.com/eager-dev/eagerx_ode/commit/475a99fa27c43b3362d818468bbe206a4b648597))


tests/test_eagerx_ode.py
# ROS packages required
import eagerx
from copy import deepcopy
import numpy as np

import pytest

NP = eagerx.NEW_PROCESS
ENV = eagerx.ENVIRONMENT


@pytest.mark.timeout(60)
@pytest.mark.parametrize(
    "eps, steps, sync, rtf, p",
    [(3, 3, True, 0, ENV)],
)
def test_ode_engine(eps, steps, sync, rtf, p):
    """
    Creates an environment with the dummy Pendulum and OdeEngine.

    :param eps: Number of episodes
    :param steps: Number of steps per episode
    :param sync: If True, the environment is reactive
    :param rtf: Real-time factor
    :param p: Process
    :return:
    """
    eagerx.set_log_level(eagerx.WARN)

    # Define unique name for test environment
    name = f"{eps}_{steps}_{sync}_{p}"
    engine_p = p
    rate = 30

    # Initialize empty graphs
    graph = eagerx.Graph.create()

    # Create pendulum
    try:
        from tests.pendulum.objects import Pendulum
        running_local = False
    except ImportError:
        from pendulum.objects import Pendulum
        running_local = True
    pendulum = Pendulum.make("pendulum")
    graph.add(pendulum)

    if running_local:
        pendulum.config.ode = "pendulum.pendulum_ode/pendulum_ode"
        pendulum.config.Dfun = "pendulum.pendulum_ode/pendulum_dfun"

    # Connect the nodes
    graph.connect(action="action", target=pendulum.actuators.pendulum_input)
    graph.connect(source=pendulum.sensors.pendulum_output, observation="observation", window=1)
    graph.connect(source=pendulum.sensors.action_applied, observation="action_applied", window=1)
    graph.connect(source=pendulum.sensors.theta, observation="theta", window=1)
    graph.connect(source=pendulum.sensors.dtheta, observation="dtheta", window=1)

    graph.render(pendulum.sensors.image, rate=10)

    # Define engines
    from eagerx_ode.engine import OdeEngine
    engine = OdeEngine.make(rate=rate, sync=sync, real_time_factor=rtf, process=engine_p)

    # Define backend
    from eagerx.backends.ros1 import Ros1
    backend = Ros1.make()
    # from eagerx.backends.single_process import SingleProcess
    # backend = SingleProcess.make()

    # Define environment
    class TestEnv(eagerx.BaseEnv):
        def __init__(self, name, rate, graph, engine, backend, force_start):
            super().__init__(name, rate, graph, engine, backend=backend, force_start=force_start)

        def step(self, action):
            obs = self._step(action)
            # Calculate reward
            if len(obs["observation"][0]) == 2:
                th, thdot = obs["observation"][0]
                sin_th = np.sin(th)
                cos_th = np.cos(th)
            else:
                sin_th, cos_th, thdot = 0, -1, 0
            th = np.arctan2(sin_th, cos_th)
            cost = th ** 2 + 0.1 * (thdot / (1 + 10 * abs(th))) ** 2
            # Determine done flag
            done = steps > 500
            # Set info:
            info = dict()
            return obs, -cost, done, info

        def reset(self):
            states = self.state_space.sample()
            obs = self._reset(states)
            return obs

    # Initialize Environment
    env = TestEnv(name, rate, graph, engine, backend, force_start=True)

    # First reset
    env.reset()
    action = env.action_space.sample()
    for j in range(eps):
        print("\n[Episode %s]" % j)
        for i in range(steps):
            obs, _, _, _ = env.step(action)
        env.reset()
    print("\n[Finished]")
    env.shutdown()
    print("\n[Shutdown]")


@pytest.mark.timeout(60)
@pytest.mark.parametrize(
    "eps, steps, sync, rtf, p",
    [(3, 30, True, 0, ENV)],
)
def test_dfun(eps, steps, sync, rtf, p):
    """
    Creates two environments, one uses a Jacobian function (Dfun) and the other not.
    Tests if the observations of the environments are close to eachother within a tolerance.

    :param eps: Number of episodes
    :param steps: Number of steps per episode
    :param sync: If True, the environment is reactive
    :param rtf: Real-time factor
    :param p: Process
    :return:
    """

    eagerx.set_log_level(eagerx.WARN)

    # Define unique name for test environment
    name = f"{eps}_{steps}_{sync}_{p}"
    engine_p = p
    rate = 30

    # Initialize empty graphs
    graph = eagerx.Graph.create()
    graph2 = eagerx.Graph.create()

    # Create pendulum
    try:
        from tests.pendulum.objects import Pendulum
        running_local = False
    except ImportError:
        from pendulum.objects import Pendulum
        running_local = True
    pendulum = Pendulum.make("pendulum")
    graph.add(pendulum)
    pendulum2 = Pendulum.make("pendulum2", Dfun=None)
    graph2.add(pendulum2)

    if running_local:
        pendulum.config.ode = "pendulum.pendulum_ode/pendulum_ode"
        pendulum.config.Dfun = "pendulum.pendulum_ode/pendulum_dfun"
        pendulum2.config.ode = "pendulum.pendulum_ode/pendulum_ode"
        pendulum2.config.Dfun = None

    # Connect the nodes
    graph.connect(action="action", target=pendulum.actuators.pendulum_input)
    graph.connect(source=pendulum.sensors.pendulum_output, observation="observation", window=1)
    graph.connect(source=pendulum.sensors.action_applied, observation="action_applied", window=1)
    graph.render(pendulum.sensors.image, rate=10)

    graph2.connect(action="action", target=pendulum2.actuators.pendulum_input)
    graph2.connect(source=pendulum2.sensors.pendulum_output, observation="observation", window=1)
    graph2.connect(source=pendulum2.sensors.action_applied, observation="action_applied", window=1)
    graph2.render(pendulum2.sensors.image, rate=10)

    # Define engines
    from eagerx_ode.engine import OdeEngine
    engine = OdeEngine.make(rate=rate, sync=sync, real_time_factor=rtf, process=engine_p)

    # Define backend
    # from eagerx.backends.ros1 import Ros1
    # backend = Ros1.make()
    from eagerx.backends.single_process import SingleProcess
    backend = SingleProcess.make()

    # Define environment
    class TestEnv(eagerx.BaseEnv):
        def __init__(self, name, rate, graph, engine, backend, force_start):
            super().__init__(name, rate, graph, engine, backend=backend, force_start=force_start)

        def step(self, action):
            obs = self._step(action)
            # Calculate reward
            if len(obs["observation"][0]) == 2:
                th, thdot = obs["observation"][0]
                sin_th = np.sin(th)
                cos_th = np.cos(th)
            else:
                sin_th, cos_th, thdot = 0, -1, 0
            th = np.arctan2(sin_th, cos_th)
            cost = th ** 2 + 0.1 * (thdot / (1 + 10 * abs(th))) ** 2
            # Determine done flag
            done = steps > 500
            # Set info:
            info = dict()
            return obs, -cost, done, info

        def reset(self):
            states = self.state_space.sample()
            for key, value in states.items():
                if "model_state" in key:
                    states[key] = value * 0.0
            obs = self._reset(states)
            return obs

    # Initialize Environment
    env = TestEnv(name, rate, graph, deepcopy(engine), backend, force_start=True)
    env2 = TestEnv(name + "_2", rate, graph2, deepcopy(engine), backend, force_start=True)

    # First reset
    obs = env.reset()
    obs2 = env2.reset()
    action = env.action_space.sample()
    for j in range(eps):
        print("\n[Episode %s]" % j)
        for i in range(steps):
            obs, _, _, _ = env.step(action)
            obs2, _, _, _ = env2.step(action)

            # Assert if result is the same with and without Jacobian
            assert np.allclose(obs["observation"], obs2["observation"])
        env.reset()
        env2.reset()
    print("\n[Finished]")
    env.shutdown()
    env2.shutdown()
    print("\n[Shutdown]")


if __name__ == "__main__":
    test_dfun(3, 30, True, 0, ENV)
    test_ode_engine(3, 3, True, 0, ENV)


tests/pendulum/objects.py
from eagerx.core.space import Space
from eagerx_ode.engine import OdeEngine
from eagerx.core.entities import Object
from eagerx.core.specs import ObjectSpec
from eagerx.core.graph_engine import EngineGraph
import eagerx.core.register as register
import numpy as np


class Pendulum(Object):
    @classmethod
    @register.sensors(
        pendulum_output=Space(low=np.array([-3.14, -9], dtype="float32"), high=np.array([3.14, 9], dtype="float32")),
        action_applied=Space(low=np.array([-3], dtype="float32"), high=np.array([3], dtype="float32")),
        image=Space(dtype="uint8"),
        theta=Space(low=-999.0, high=999.0, shape=(), dtype="float32"),
        dtheta=Space(low=-999.0, high=999.0, shape=(), dtype="float32")
    )
    @register.actuators(pendulum_input=Space(low=np.array([-3], dtype="float32"), high=np.array([3], dtype="float32")))
    @register.engine_states(model_state=Space(low=np.array([-3.14, -9], dtype="float32"), high=np.array([3.14, 9], dtype="float32")),
                            model_parameters=Space(dtype="float32"))
    def make(
        cls,
        name: str,
        sensors=None,
        states=None,
        rate=30,
        ode="tests.pendulum.pendulum_ode/pendulum_ode",
        Dfun="tests.pendulum.pendulum_ode/pendulum_dfun",
    ):
        """Object spec of pendulum"""
        spec = cls.get_specification()

        # Modify default agnostic params
        # Only allow changes to the agnostic params (rates, windows, (space)converters, etc...
        spec.config.name = name
        spec.config.sensors = sensors if sensors else ["pendulum_output", "action_applied", "image", "theta", "dtheta"]
        spec.config.actuators = ["pendulum_input"]
        spec.config.states = states if states else ["model_state"]

        # Add registered agnostic params
        spec.config.render_shape = [480, 480]
        spec.config.ode = ode
        spec.config.Dfun = Dfun

        # Set observation properties: (space_converters, rate, etc...)
        spec.sensors.pendulum_output.rate = rate
        spec.sensors.action_applied.rate = rate
        spec.sensors.dtheta.rate = rate
        spec.sensors.theta.rate = rate
        spec.actuators.pendulum_input.rate = rate
        spec.sensors.image.rate = 15

        # Set image space
        shape = (spec.config.render_shape[0], spec.config.render_shape[1], 3)
        spec.sensors.image.space = Space(low=0, high=255, shape=shape, dtype="uint8")

        # Set model_parameters properties: (space_converters) # [J, m, l, b0, K, R, c, a]
        fixed = [0.000189238, 0.0563641, 0.0437891, 0.000142205, 0.0502769, 9.83536]
        diff = [0, 0, 0, 0.00, 0.00]  # Percentual delta with respect to fixed value
        low = np.array([val - diff * val for val, diff in zip(fixed, diff)], dtype="float32")
        high = np.array([val + diff * val for val, diff in zip(fixed, diff)], dtype="float32")
        spec.states.model_parameters.space = Space(low=low, high=high)

        return spec

    @staticmethod
    @register.engine(OdeEngine)  # This decorator pre-initializes engine implementation with default object_params
    def ode_engine(spec: ObjectSpec, graph: EngineGraph):
        """Engine-specific implementation (OdeEngine) of the object."""
        # Set object arguments (nothing to set here in this case)
        spec.engine.ode = spec.config.ode
        # Set default params of pendulum ode [J, m, l, b0, K, R].
        spec.engine.ode_params = [
            0.000189238,
            0.0563641,
            0.0437891,
            0.000142205,
            0.0502769,
            9.83536,
        ]
        spec.engine.Dfun = spec.config.Dfun

        # Create engine states (no agnostic states defined in this case)
        from eagerx_ode.engine_states import OdeEngineState, OdeParameters
        spec.engine.states.model_state = OdeEngineState.make()
        spec.engine.states.model_parameters = OdeParameters.make(list(range(5)))

        # Create sensor engine nodes
        from eagerx_ode.engine_nodes import OdeOutput, OdeInput, OdeFloatOutput, OdeRender, ActionApplied
        obs = OdeOutput.make("pendulum_output", rate=spec.sensors.pendulum_output.rate, process=2)
        image = OdeRender.make(
            "image",
            shape=spec.config.render_shape,
            render_fn="tests.pendulum.pendulum_render/pendulum_render_fn",
            rate=spec.sensors.image.rate,
            process=0,
        )
        theta = OdeFloatOutput.make("theta", rate=spec.sensors.theta.rate, process=2, idx=0)

        dtheta = OdeFloatOutput.make("dtheta", rate=spec.sensors.dtheta.rate, process=2, idx=1)

        # Create actuator engine nodes
        action = OdeInput.make(
            "pendulum_actuator",
            rate=spec.actuators.pendulum_input.rate,
            process=2,
            default_action=[0],
        )

        # Connect all engine nodes
        graph.add([obs, image, action, theta, dtheta])
        graph.connect(source=obs.outputs.observation, sensor="pendulum_output")
        graph.connect(actuator="pendulum_input", target=action.inputs.action)
        graph.connect(
            source=action.outputs.action_applied,
            target=image.inputs.action_applied,
            skip=True,
        )
        graph.connect(source=obs.outputs.observation, target=image.inputs.observation)
        graph.connect(source=image.outputs.image, sensor="image")

        # Add action applied
        applied = ActionApplied.make("applied", rate=spec.sensors.action_applied.rate, process=2)
        graph.add(applied)
        graph.connect(
            source=action.outputs.action_applied,
            target=applied.inputs.action_applied,
            skip=True,
        )
        graph.connect(source=applied.outputs.action_applied, sensor="action_applied")
        graph.connect(source=theta.outputs.observation, sensor="theta")
        graph.connect(source=dtheta.outputs.observation, sensor="dtheta")


tests/pendulum/pendulum_ode.py
from math import sin, cos


def pendulum_ode(x, t, u, J, m, l, b, K, R):  # noqa:
    g = 9.81

    ddx = (m * g * l * sin(x[0]) - x[1] * (b + K * K / R) + K * u / R) / J

    return [x[1], ddx]


def pendulum_dfun(x, t, u, J, m, l, b, K, R):  # noqa:
    g = 9.81

    return [[0, 1], [m * g * l * cos(x[0]) / J, - (b + K * K / R) / J]]


eagerx_ode/bridge.py
# OTHER
from typing import Optional, Dict, Union, List
from scipy.integrate import odeint

# ROS IMPORTS
import rospy
from std_msgs.msg import UInt64
from genpy.message import Message

# RX IMPORTS
from eagerx.core.constants import process, ERROR
import eagerx.core.register as register
from eagerx.core.entities import Bridge
from eagerx.core.specs import BridgeSpec
from eagerx.utils.utils import Msg, get_attribute_from_module


class OdeBridge(Bridge):
    @staticmethod
    @register.spec("OdeBridge", Bridge)
    def spec(
        spec: BridgeSpec,
        rate,
        sync: Optional[bool] = True,
        process: Optional[int] = process.ENVIRONMENT,
        real_time_factor: Optional[float] = 0,
        simulate_delays: Optional[bool] = True,
        log_level: Optional[int] = ERROR,
        rtol: float = 2e-8,
        atol: float = 2e-8,
        hmax: float = 0.0,
        hmin: float = 0.0,
        mxstep: int = 0,
    ):
        """
        Spec of the OdeBridge

        :param spec: Not provided by the user.
        :param rate: Rate of the bridge
        :param process: {0: NEW_PROCESS, 1: ENVIRONMENT, 2: BRIDGE, 3: EXTERNAL}
        :param sync: Run reactive or async
        :param real_time_factor: simulation speed. 0 == "as fast as possible".
        :param simulate_delays: Boolean flag to simulate delays.
        :param log_level: {0: SILENT, 10: DEBUG, 20: INFO, 30: WARN, 40: ERROR, 50: FATAL}
        :param rtol: The input parameters rtol and atol determine the error control performed by the solver.
        :param atol: The input parameters rtol and atol determine the error control performed by the solver.
        :param hmax: The maximum absolute step size allowed.
        :param hmin: The minimum absolute step size allowed.
        :param mxstep: Maximum number of (internally defined) steps allowed for each integration point in t.
        :return: BridgeSpec
        """
        # Modify default bridge params
        spec.config.rate = rate
        spec.config.process = process
        spec.config.sync = sync
        spec.config.real_time_factor = real_time_factor
        spec.config.simulate_delays = simulate_delays
        spec.config.log_level = log_level
        spec.config.color = "magenta"

        # Add custom params
        custom = dict(rtol=rtol, atol=atol, hmax=hmax, hmin=hmin, mxstep=mxstep)
        spec.config.update(custom)

    def initialize(self, rtol, atol, hmax, hmin, mxstep):
        # Initialize any simulator here, that is passed as reference to each engine node
        self.odeint_args = dict(rtol=rtol, atol=atol, hmax=hmax, hmin=hmin, mxstep=mxstep)
        self.simulator = dict()

    @register.bridge_config(ode=None, Dfun=None, ode_params=list())
    def add_object(self, config, bridge_config, node_params, state_params):
        # add object to simulator (we have a ref to the simulator with self.simulator)
        rospy.loginfo(f'Adding object "{config["name"]}" of type "{config["entity_id"]}" to the simulator.')

        # Extract relevant agnostic params
        obj_name = config["name"]
        ode = get_attribute_from_module(bridge_config["ode"])
        Dfun = get_attribute_from_module(bridge_config["Dfun"]) if "Dfun" in config and config["Dfun"] else None

        # Create new env, and add to simulator
        self.simulator[obj_name] = dict(
            ode=ode,
            Dfun=Dfun,
            state=None,
            input=None,
            ode_params=bridge_config["ode_params"],
        )

    def pre_reset(self, **kwargs: Optional[Msg]):
        pass

    @register.states()
    def reset(self, **kwargs: Optional[Msg]):
        pass

    @register.outputs(tick=UInt64)
    def callback(self, t_n: float, **kwargs: Dict[str, Union[List[Message], float, int]]):
        for _obj_name, sim in self.simulator.items():
            input = sim["input"]
            ode = sim["ode"]
            Dfun = sim["Dfun"]
            x = sim["state"]
            ode_params = sim["ode_params"]
            if x is not None and input is not None:
                sim["state"] = odeint(
                    ode,
                    x,
                    [0, 1.0 / self.rate],
                    args=(input, *ode_params),
                    Dfun=Dfun,
                    **self.odeint_args,
                )[-1]


eagerx_ode/engine.py
# OTHER
from typing import Optional, List
from scipy.integrate import odeint

# RX IMPORTS
from eagerx.core.constants import process, ERROR
import eagerx.core.register as register
from eagerx.core.entities import Engine
from eagerx.utils.utils import load


class OdeEngine(Engine):
    @classmethod
    def make(
        cls,
        rate: float,
        sync: Optional[bool] = True,
        process: Optional[int] = process.ENVIRONMENT,
        real_time_factor: Optional[float] = 0,
        simulate_delays: Optional[bool] = True,
        log_level: Optional[int] = ERROR,
        rtol: float = 2e-8,
        atol: float = 2e-8,
        hmax: float = 0.0,
        hmin: float = 0.0,
        mxstep: int = 0,
    ):
        """
        Spec of the OdeEngine

        :param rate: Rate of the engine
        :param process: {0: NEW_PROCESS, 1: ENVIRONMENT, 2: ENGINE, 3: EXTERNAL}
        :param sync: Run reactive or async
        :param real_time_factor: simulation speed. 0 == "as fast as possible".
        :param simulate_delays: Boolean flag to simulate delays.
        :param log_level: {0: SILENT, 10: DEBUG, 20: INFO, 30: WARN, 40: ERROR, 50: FATAL}
        :param rtol: The input parameters rtol and atol determine the error control performed by the solver.
        :param atol: The input parameters rtol and atol determine the error control performed by the solver.
        :param hmax: The maximum absolute step size allowed.
        :param hmin: The minimum absolute step size allowed.
        :param mxstep: Maximum number of (internally defined) steps allowed for each integration point in t.
        :return: EngineSpec
        """
        spec = cls.get_specification()

        # Modify default engine params
        spec.config.rate = rate
        spec.config.process = process
        spec.config.sync = sync
        spec.config.real_time_factor = real_time_factor
        spec.config.simulate_delays = simulate_delays
        spec.config.log_level = log_level
        spec.config.color = "magenta"

        # Add custom params
        custom = dict(rtol=rtol, atol=atol, hmax=hmax, hmin=hmin, mxstep=mxstep)
        spec.config.update(custom)
        return spec

    def initialize(self, spec):
        rtol = spec.config.rtol
        atol = spec.config.atol
        hmax = spec.config.hmax
        hmin = spec.config.hmin
        mxstep = spec.config.mxstep

        # Initialize any simulator here, that is passed as reference to each engine node
        self.odeint_args = dict(rtol=rtol, atol=atol, hmax=hmax, hmin=hmin, mxstep=mxstep)
        self.simulator = dict()

    def add_object(self, name, ode: str = None, Dfun: Optional[str] = None, ode_params: Optional[List] = None):
        # add object to simulator (we have a ref to the simulator with self.simulator)
        self.backend.loginfo(f'Adding object "{name}" to the simulator.')

        # Extract relevant agnostic params
        ode = load(ode)
        Dfun = load(Dfun) if Dfun is not None else None

        # Create new env, and add to simulator
        self.simulator[name].update(
            ode=ode,
            Dfun=Dfun,
            state=None,
            input=None,
            ode_params=ode_params,
        )

    def pre_reset(self):
        pass

    @register.states()
    def reset(self):
        pass

    def callback(self, t_n: float):
        for _obj_name, sim in self.simulator.items():
            input = sim["input"]
            ode = sim["ode"]
            Dfun = sim["Dfun"]
            x = sim["state"]
            ode_params = sim["ode_params"]
            if x is not None and input is not None:
                sim["state"] = odeint(
                    ode,
                    x,
                    [0, 1.0 / self.rate],
                    args=(input, *ode_params),
                    Dfun=Dfun,
                    **self.odeint_args,
                )[-1]


eagerx_ode/enginestates.py
import numpy as np
from eagerx.core.entities import EngineState
import eagerx.core.register as register


class OdeEngineState(EngineState):
    @staticmethod
    @register.spec("OdeSimState", EngineState)
    def spec(spec):
        spec.initialize(OdeEngineState)

    def initialize(self):
        self.obj_name = self.agnostic_params["name"]

    def reset(self, state, done):
        self.simulator[self.obj_name]["state"] = np.squeeze(state.data)


class OdeParameters(EngineState):
    @staticmethod
    @register.spec("OdeParameters", EngineState)
    def spec(spec, indices):
        spec.initialize(OdeParameters)
        spec.agnostic_params.indices = indices

    def initialize(self, indices):
        self.obj_name = self.agnostic_params["name"]
        self.indices = indices

    def reset(self, state, done):
        for i in self.indices:
            self.simulator[self.obj_name]["ode_params"][i] = state.data[i]


eagerx_ode/pendulum/objects.py
# ROS IMPORTS
from std_msgs.msg import Float32MultiArray

# EAGERx IMPORTS
from eagerx_ode.bridge import OdeBridge
from eagerx.core.entities import (
    Object,
    EngineNode,
    SpaceConverter,
    EngineState,
)
from eagerx.core.specs import ObjectSpec
from eagerx.core.graph_engine import EngineGraph
import eagerx.core.register as register


class Pendulum(Object):
    entity_id = "Eagerx_Ode_Pendulum"

    @staticmethod
    @register.sensors(pendulum_output=Float32MultiArray, action_applied=Float32MultiArray)
    @register.actuators(pendulum_input=Float32MultiArray)
    @register.engine_states(model_state=Float32MultiArray, model_parameters=Float32MultiArray)
    @register.config()
    def agnostic(spec: ObjectSpec, rate):
        """Agnostic definition of the pendulum"""
        # Register standard converters, space_converters, and processors
        import eagerx.converters  # noqa # pylint: disable=unused-import

        # Set observation properties: (space_converters, rate, etc...)
        spec.sensors.pendulum_output.rate = rate
        spec.sensors.pendulum_output.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", low=[-3.14, -9], high=[3.14, 9], dtype="float32"
        )

        spec.sensors.action_applied.rate = rate
        spec.sensors.action_applied.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", low=[-3], high=[3], dtype="float32"
        )

        # Set actuator properties: (space_converters, rate, etc...)
        spec.actuators.pendulum_input.rate = rate
        spec.actuators.pendulum_input.window = 1
        spec.actuators.pendulum_input.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", low=[-3], high=[3], dtype="float32"
        )

        # Set model_state properties: (space_converters)
        spec.states.model_state.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray",
            low=[-3.14159265359, -9],
            high=[3.14159265359, 9],
            dtype="float32",
        )

        # Set model_parameters properties: (space_converters) # [J, m, l, b0, K, R, c, a]
        fixed = [0.000189238, 0.0563641, 0.0437891, 0.000142205, 0.0502769, 9.83536]
        diff = [0, 0, 0, 0.05, 0.05]  # Percentual delta with respect to fixed value
        low = [val - diff * val for val, diff in zip(fixed, diff)]
        high = [val + diff * val for val, diff in zip(fixed, diff)]
        spec.states.model_parameters.space_converter = SpaceConverter.make(
            "Space_Float32MultiArray", low=low, high=high, dtype="float32"
        )

    @staticmethod
    @register.spec(entity_id, Object)
    def spec(spec: ObjectSpec, name: str, sensors=None, states=None, rate=30):
        """Object spec of pendulum"""
        # Performs all the steps to fill-in the params with registered info about all functions.
        Pendulum.initialize_spec(spec)

        # Modify default agnostic params
        # Only allow changes to the agnostic params (rates, windows, (space)converters, etc...
        spec.config.name = name
        spec.config.sensors = sensors if sensors else ["pendulum_output", "action_applied"]
        spec.config.actuators = ["pendulum_input"]
        spec.config.states = states if states else ["model_state"]

        # Add bridge implementation
        Pendulum.agnostic(spec, rate)

    @staticmethod
    @register.bridge(entity_id, OdeBridge)  # This decorator pre-initializes bridge implementation with default object_params
    def ode_bridge(spec: ObjectSpec, graph: EngineGraph):
        """Engine-specific implementation (OdeBridge) of the object."""
        # Import any object specific entities for this bridge
        import eagerx_ode.pendulum.pendulum_ode  # noqa # pylint: disable=unused-import

        # Set object arguments (nothing to set here in this case)
        spec.OdeBridge.ode = "eagerx_ode.pendulum.pendulum_ode/pendulum_ode"
        # Set default params of pendulum ode [J, m, l, b0, K, R].
        spec.OdeBridge.ode_params = [
            0.000189238,
            0.0563641,
            0.0437891,
            0.000142205,
            0.0502769,
            9.83536,
        ]

        # Create engine states (no agnostic states defined in this case)
        spec.OdeBridge.states.model_state = EngineState.make("OdeEngineState")
        spec.OdeBridge.states.model_parameters = EngineState.make("OdeParameters", list(range(5)))

        # Create sensor engine nodes
        obs = EngineNode.make(
            "OdeOutput",
            "pendulum_output",
            rate=spec.sensors.pendulum_output.rate,
            process=2,
        )

        # Create actuator engine nodes
        action = EngineNode.make(
            "OdeInput",
            "pendulum_actuator",
            rate=spec.actuators.pendulum_input.rate,
            process=2,
            default_action=[0],
        )

        # Connect all engine nodes
        graph.add([obs, action])
        graph.connect(source=obs.outputs.observation, sensor="pendulum_output")
        graph.connect(actuator="pendulum_input", target=action.inputs.action)

        # Add action applied
        applied = EngineNode.make("ActionApplied", "applied", rate=spec.sensors.action_applied.rate, process=2)
        graph.add(applied)
        graph.connect(
            source=action.outputs.action_applied,
            target=applied.inputs.action_applied,
            skip=True,
        )
        graph.connect(source=applied.outputs.action_applied, sensor="action_applied")
